internal/command/deploy/plan.go

```diff
@@ -0,0 +1,691 @@
+
+			currentState, err := md.appState(ctx, oldAppState)
+			if err != nil {
+				span.RecordError(updateErr)
+				return fmt.Errorf("failed to get current app state: %w", err)
+			}
+			err = md.updateMachinesWRecovery(ctx, currentState, newAppState, sl, updateMachineSettings{
+				pushForward:          false,
+				skipHealthChecks:     settings.skipHealthChecks,
+				skipSmokeChecks:      settings.skipSmokeChecks,
+				skipLeaseAcquisition: true,
+			})
+			if err == nil {
+				break
+			} else if errors.Is(err, context.Canceled) {
+				span.RecordError(updateErr)
+				return err
+			} else {
+				if errors.As(err, &unrecoverableErr) {
+					span.RecordError(updateErr)
+					return err
+				}
+				fmt.Fprintln(md.io.ErrOut, "Failed to update machines:", err, "Retrying...")
+			}
+			attempts += 1
+			time.Sleep(1 * time.Second)
+		}
+	}
+
+	return nil
+}
+
+func (md *machineDeployment) updateProcessGroup(ctx context.Context, machineTuples []machinePairing, statusLines map[string]statuslogger.StatusLine, poolSize int) error {
+	ctx, span := tracing.GetTracer().Start(ctx, "update_process_group")
+	defer span.End()
+
+	group := errgroup.Group{}
+	group.SetLimit(poolSize)
+
+	for _, machPair := range machineTuples {
+		machPair := machPair
+		oldMachine := machPair.oldMachine
+		newMachine := machPair.newMachine
+
+		group.Go(func() error {
+			checkResult, _ := healthChecksPassed.Load(machPair.oldMachine.ID)
+			machineCheckResult := checkResult.(*healthcheckResult)
+
+			var sl statuslogger.StatusLine
+			if oldMachine != nil {
+				sl = statusLines[oldMachine.ID]
+			} else if newMachine != nil {
+				sl = statusLines[newMachine.ID]
+			}
+
+			err := md.updateMachineWChecks(ctx, oldMachine, newMachine, sl, md.io, machineCheckResult)
+			if err != nil {
+				sl.LogStatus(statuslogger.StatusFailure, err.Error())
+				span.RecordError(err)
+				return fmt.Errorf("failed to update machine %s: %w", oldMachine.ID, err)
+			}
+			return nil
+		})
+	}
+
+	if err := group.Wait(); err != nil {
+		span.RecordError(err)
+		return err
+	}
+
+	return nil
+}
+
+func (md *machineDeployment) acquireLeases(ctx context.Context, machineTuples []machinePairing, poolSize int, statusLogger statuslogger.StatusLogger) error {
+	ctx, span := tracing.GetTracer().Start(ctx, "acquire_leases")
+
+	leaseGroup := errgroup.Group{}
+	leaseGroup.SetLimit(poolSize)
+
+	for idx, machineTuple := range machineTuples {
+		machineTuple := machineTuple
+		idx := idx
+
+		leaseGroup.Go(func() error {
+			sl := statusLogger.Line(idx)
+
+			var machine *fly.Machine
+			if machineTuple.oldMachine != nil {
+				machine = machineTuple.oldMachine
+			} else if machineTuple.newMachine != nil {
+				machine = machineTuple.newMachine
+			} else {
+				return nil
+			}
+			sl.LogStatus(statuslogger.StatusRunning, fmt.Sprintf("Acquiring lease for %s", machine.ID))
+
+			if machine.LeaseNonce == "" {
+				sl.LogStatus(statuslogger.StatusRunning, fmt.Sprintf("Waiting for job %s", machine.ID))
+			}
```

Comment:
Do you remember why we show `Waiting for job` in this case? It seems like we should still say `Acquiring lease for` if the machine isn't current leased, right?

Reply:
that's a great question, this seems like a bug.  I feel like we should check if LeaseNonce != "", and if that's the case, we just return (since we already have an acquired lease)

Reply:
The lease nonce could be set if someone else has it too though, which which case we still want to wait. I'm going to remove this LogStatus though.